src/libtoast/src/toast_tod_filter.cpp

```diff
@@ -367,3 +367,120 @@ void toast::add_templates(double * signal, double * templates, double * coeff,
+                //            =  [[a_1 * s_1 * m_1 + a_2 * s_2 * m_2],
+                //                [b_1 * s_1 * m_1 + b_2 * s_2 * m_2],
+                //                [c_1 * s_1 * m_1 + c_2 * s_2 * m_2]]
+                //
+                //          A = (mask * templates)^T  X  (mask * templates)
+                //            = [ [a_1 * a_1 * m_1 + a_2 * a_2 * m_2,
+                //                 a_1 * b_1 * m_1 + a_2 * b_2 * m_2,
+                //                 a_1 * c_1 * m_1 + a_2 * c_2 * m_2],
+                //                [b_1 * a_1 * m_1 + b_2 * a_2 * m_2,
+                //                 b_1 * b_1 * m_1 + b_2 * b_2 * m_2,
+                //                 b_1 * c_1 * m_1 + b_2 * c_2 * m_2],
+                //                [c_1 * a_1 * m_1 + c_2 * a_2 * m_2,
+                //                 c_1 * b_1 * m_1 + c_2 * b_2 * m_2,
+                //                 c_1 * c_1 * m_1 + c_2 * c_2 * m_2] ]
+                //
+
+                for (int32_t idet = 0; idet < ndet; ++idet) {
+                    // For each detector...
+                    if (det_group[idet] != igroup) {
+                        // This detectors is not in this group
+                        continue;
+                    }
+
+                    // Mask value for this detector
+                    double det_mask = (masks[isamp * ndet + idet] == 0) ? 0.0 : 1.0;
+
+                    // Signal value for this detector
+                    double det_sig = signals[isamp * ndet + idet];
+
+                    for (int32_t imode = 0; imode < nmode; ++imode) {
+                        int32_t tmpl_off = idet * nmode;
+                        rhs[imode] += templates[tmpl_off + imode] * det_sig * det_mask;
+
+                        for (int32_t jmode = imode; jmode < nmode; ++jmode) {
+                            double val = templates[tmpl_off + imode] * templates[tmpl_off + jmode] * det_mask;
+                            A[imode * nmode + jmode] += val;
+                            if (jmode > imode) {
+                                A[jmode * nmode + imode] += val;
+                            }
+                        }
+                    }
+                }
+
+                // DGELSS will overwrite RHS with the fitting
+                // coefficients.  A is overwritten with
+                // singular vectors.
+                toast::lapack_dgelss(
+                    &inmode, &inmode, &one, A.data(), &inmode,
+                    rhs.data(), &inmode, singular_values.data(), &rcond_limit,
+                    &rank, work.data(), &lwork, &info
+                );
+                int64_t offset = isamp * (ngroup * nmode) + igroup * nmode;
+                if (info == 0) {
+                    // Solve was successful
+                    for (int64_t m = 0; m < nmode; ++m) {
+                        coeff[offset + m] = rhs[m];
+                    }
+                } else {
```

Comment:
I wonder under what circumstances the minimum norm solution could not be found. I'm uncomfortable with failing silently and writing out zeros. Seems like corresponding detector samples should be flagged. I realize that the Python implementation did not address this either since the least squares solution is pretty much guaranteed to be found.

Reply:
But the detector samples are already flagged if the coefficients are zero: